### PR TITLE
Clean up configure.ac

### PR DESCRIPTION
**Pull request by boehms**
- Add check for -wrap in linker
- Add checks for wrapped functions
- make gotcha optional
- remove leveldb and flatcc from the optional dependencies

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the UnifyFS code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted.

### CODE REVIEW

File: client/src/unifyfs-sysio.c [hunk 1]
```diff
@@ -40,6 +40,10 @@
  * Please also read this file LICENSE.CRUISE
  */
 
+#include <sys/stat.h>
```
Comment (MichaelBrim) [inline client/src/unifyfs-sysio.c]: sys/stat and config.h are already included by unifyfs_internal.h

File: m4/flatcc.m4 [hunk 2]
```diff
@@ -3,21 +3,21 @@ AC_DEFUN([UNIFYFS_AC_FLATCC], [
   FLATCC_OLD_CFLAGS=$CFLAGS
   FLATCC_OLD_LDFLAGS=$LDFLAGS
 
-  AC_ARG_WITH([flatcc], [AC_HELP_STRING([--with-flatcc=PATH],
-    [path to installed libflatcc [default=/usr/local]])], [
-    FLATCC_CFLAGS="-I${withval}/include"
-    FLATCC_LDFLAGS="-L${withval}/lib"
+  if [[ ! -z $FLATCC_ROOT ]]; then
```
Comment (MichaelBrim) [inline m4/flatcc.m4]: shouldn't use bash [[ ]] in m4 code. `"! -z" == "-n"`

File: m4/leveldb.m4 [hunk 3]
```diff
@@ -3,21 +3,21 @@ AC_DEFUN([UNIFYFS_AC_LEVELDB], [
   LEVELDB_OLD_CFLAGS=$CFLAGS
   LEVELDB_OLD_LDFLAGS=$LDFLAGS
 
-  AC_ARG_WITH([leveldb], [AC_HELP_STRING([--with-leveldb=PATH],
-    [path to installed libleveldb [default=/usr/local]])], [
-    LEVELDB_CFLAGS="-I${withval}/include"
-    LEVELDB_LDFLAGS="-L${withval}/lib -L${withval}/lib64"
+  if [[ ! -z $LEVELDB_ROOT ]]; then
```
Comment (MichaelBrim) [inline m4/leveldb.m4]: same comment as above

File: client/src/unifyfs-sysio.c (hunk 1 above)
Comment (boehms) [reply client/src/unifyfs-sysio.c]: fixed

File: m4/flatcc.m4 [hunk 4]
```diff
@@ -3,21 +3,21 @@ AC_DEFUN([UNIFYFS_AC_FLATCC], [
   FLATCC_OLD_CFLAGS=$CFLAGS
   FLATCC_OLD_LDFLAGS=$LDFLAGS
 
-  AC_ARG_WITH([flatcc], [AC_HELP_STRING([--with-flatcc=PATH],
-    [path to installed libflatcc [default=/usr/local]])], [
-    FLATCC_CFLAGS="-I${withval}/include"
-    FLATCC_LDFLAGS="-L${withval}/lib"
+  if test -n $FLATCC_ROOT ; then
```
Comment (MichaelBrim) [inline m4/flatcc.m4]: you need quotes around the variable when not in [[ ]]

File: m4/leveldb.m4 [hunk 5]
```diff
@@ -3,21 +3,21 @@ AC_DEFUN([UNIFYFS_AC_LEVELDB], [
   LEVELDB_OLD_CFLAGS=$CFLAGS
   LEVELDB_OLD_LDFLAGS=$LDFLAGS
 
-  AC_ARG_WITH([leveldb], [AC_HELP_STRING([--with-leveldb=PATH],
-    [path to installed libleveldb [default=/usr/local]])], [
-    LEVELDB_CFLAGS="-I${withval}/include"
-    LEVELDB_LDFLAGS="-L${withval}/lib -L${withval}/lib64"
+  if test -n $LEVELDB_ROOT ; then
```
Comment (MichaelBrim) [inline m4/leveldb.m4]: need quotes around the variable

File: configure.ac [hunk 6]
```diff
@@ -196,7 +208,13 @@ AM_CONDITIONAL([HAVE_HDF5], [test x$with_hdf5 = xyes])
...
 CP_WRAPPERS+=",-wrap,chmod"
 CP_WRAPPERS+=",-wrap,fchmod"
-CP_WRAPPERS+=",-wrap,lio_listio"
+
+OLD_LIBS=$LIBS
+LIBS+="-lrt"
+AC_CHECK_FUNCS(lio_listio,[
+    CP_WRAPPERS+=",-wrap,lio_listio"
+], [])
+LIBS=$OLDLIBS
```
Comment (MichaelBrim) [inline configure.ac]: s/OLDLIBS/OLD_LIBS/

File: configure.ac [hunk 7]
```diff
@@ -79,69 +81,85 @@ AC_CHECK_FUNCS([gethostbyname strcasecmp strdup strerror strncasecmp strrchr])
...
-AC_ARG_ENABLE(st-dev-workaround, 
-[  --enable-st-dev-workaround Gather device id from parent directory instead of file], 
-[if test "x$enableval" = "xyes" ; then
-    AC_DEFINE(__CP_ST_DEV_WORKAROUND, 1, Define if device id should be taken from parent directory rather than file)
-    DISABLE_LDPRELOAD="1"
-fi]
-,)
+# look for MPI and set flags
+LX_FIND_MPI
+#AS_IF([test "x$enable_fortran" = "xyes"],[
```
Comment (MichaelBrim) [inline configure.ac]: remove commented line

File: configure.ac [hunk 8]
```diff
@@ -79,69 +81,85 @@ AC_CHECK_FUNCS([gethostbyname strcasecmp strdup strerror strncasecmp strrchr])
...
+if test "x$enable_fortran" = "xyes" ; then
+    AC_LANG_PUSH([Fortran])
+    LX_FIND_MPI
+    AC_LANG_POP
+#],[])
```
Comment (MichaelBrim) [inline configure.ac]: remove commented line

File: configure.ac [hunk 9]
```diff
@@ -66,8 +68,8 @@ AC_CHECK_HEADERS([fcntl.h limits.h stdlib.h string.h sys/socket.h sys/time.h])
 AC_CHECK_HEADERS([unistd.h arpa/inet.h inttypes.h netdb.h netinet/in.h])
 AC_CHECK_HEADERS([stddef.h stdint.h libgen.h strings.h syslog.h])
 AC_CHECK_HEADERS([inttypes.h wchar.h wctype.h])
-AC_CHECK_HEADER([openssl/md5.h], [], [AC_MSG_FAILURE([
-        *** openssl/md5.h missing, openssl-devel package required])])
+dnl AC_CHECK_HEADER([openssl/md5.h], [], [AC_MSG_FAILURE([
```
Comment (MichaelBrim) [inline configure.ac]: wip? if so, remove until it's ready

File: configure.ac [hunk 10]
```diff
@@ -16,32 +19,28 @@ AM_SILENT_RULES([yes])
...
-AC_FC_LIBRARY_LDFLAGS
-AC_FC_DUMMY_MAIN
-AM_CONDITIONAL([HAVE_FORTRAN], [true])
-else
-AC_MSG_RESULT(no)
-AM_CONDITIONAL([HAVE_FORTRAN], [false])
-fi
+AS_IF([test "x$enable_fortran" = "xyes"],[
+    AC_MSG_RESULT(yes)
+    #UNIFY_SETUP_FC()
```
Comment (MichaelBrim) [inline configure.ac]: what's this?

**Comment by adammoody** [comment]
Thanks for all of the work on this @boehms and @MichaelBrim !

**Comment by adammoody** [comment]
@boehms , so my configure steps changed from
```
export PKG_CONFIG_PATH=/path/to/install/lib/pkgconfig
./autogen.sh && ./configure CFLAGS="-g -O0" \
--with-leveldb=/path/to/install \
--with-gotcha=/path/to/install \
--with-flatcc=/path/to/install \
--prefix=$installdir
```
to
```
export PKG_CONFIG_PATH=/path/to/install/lib/pkgconfig
export LEVELDB_ROOT=/path/to/install
export FLATCC_ROOT=/path/to/install
./autogen.sh && ./configure CFLAGS="-g -O0" \
--with-gotcha=/path/to/install \
--prefix=$installdir
```
Does that agree with what you'd expect?

If so we need to change our documentation to reflect this, too.

**Comment by boehms** [comment]
@adammoody That works, another option is:
```
./autogen.sh && ./configure \
--with-gotcha=/path/to/install \
--prefix=$installdir \
PKG_CONFIG_PATH=/path/to/install/lib/pkgconfig \
LEVELDB_ROOT=/path/to/install \
FLATCC_ROOT=/path/to/install
```

**Pull request merged**